### PR TITLE
ddl: avoid assert in hook function.

### PR DESCRIPTION
**Pull request by coocood**
Hook function runs in another goroutine, assert failed will silently exit that goroutine.

### CODE REVIEW

**Comment by coocood** [comment]
@zimulala @tiancaiamao PTAL

File: ddl/foreign_key_test.go [hunk 1]
```diff
@@ -137,19 +134,26 @@ func (s *testForeighKeySuite) TestForeignKey(c *C) {
...
 		mu.Lock()
+		defer mu.Unlock()
+		ti, err := testGetTableInfo(d, s.dbInfo.ID, tblInfo.ID)
+		if err != nil {
+			hookErr = errors.Trace(err)
+			return
+		}
+		fk := getForeignKey(ti, "c1_fk")
+		if fk == nil {
+			hookErr = errors.New("foreign key not exists")
```
Comment (tiancaiamao) [inline ddl/foreign_key_test.go]: hookErr never userd?  
call `c.Fail()` here?

Comment (zimulala) [inline ddl/foreign_key_test.go]: `hookErr` is checked in line167.

File: ddl/column_test.go [hunk 2]
```diff
@@ -813,19 +840,24 @@ func (s *testColumnSuite) TestDropColumn(c *C) {
...
 	var mu sync.Mutex
 
 	tc := &testDDLCallback{}
 	tc.onJobUpdated = func(job *model.Job) {
 		if checkOK {
 			return
 		}
-		t := testGetTable(c, d, s.dbInfo.ID, tblInfo.ID).(*tables.Table)
-		col := table.FindCol(t.Columns, colName)
+		mu.Lock()
```
Comment (zimulala) [inline ddl/column_test.go]: Move this line to line851. `checkOK` needs to be locked.

**Comment by zimulala** [comment]
LGTM

File: ddl/column_test.go [hunk 3]
```diff
@@ -774,8 +797,12 @@ func (s *testColumnSuite) TestAddColumn(c *C) {
 	d.start()
 
 	job := testCreateColumn(c, ctx, d, s.dbInfo, tblInfo, newColName, &ast.ColumnPosition{Tp: ast.ColumnPositionNone}, defaultColValue)
-	c.Assert(errors.ErrorStack(checkErr), Equals, "")
+
 	testCheckJobDone(c, d, job, true)
+	mu.Lock()
+	c.Assert(errors.ErrorStack(hookErr), Equals, "")
```
Comment (tiancaiamao) [inline ddl/column_test.go]: if Assert panic here, `mu.Unlock` would never be called, does it matter?

File: ddl/column_test.go [hunk 4]
```diff
@@ -740,24 +754,33 @@ func (s *testColumnSuite) TestAddColumn(c *C) {
...
 	defaultColValue := int64(4)
+
+	var mu sync.Mutex
+	var hookErr error
 	checkOK := false
 
 	tc := &testDDLCallback{}
-	var checkErr error
 	tc.onJobUpdated = func(job *model.Job) {
+		mu.Lock()
```
Comment (tiancaiamao) [inline ddl/column_test.go]: Will onJobUpdated run parallelly?  why it need a mutex

Comment (coocood) [inline ddl/column_test.go]: We need to access `hookError` inside and outside hook function. hook function is run in different goroutine.

File: ddl/column_test.go (hunk 3 above)
Comment (coocood) [inline ddl/column_test.go]: I tested it, it does has problem.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**